[SPARK-41103][CONNECT][DOC] Document how to add a new proto field of messages (#38605)
apache/spark: connector/connect/docs/adding-proto-messages.md

# Required, Optional and default values

> Connect adopts proto3, which does not support `required` constraint anymore. 

Reviewer: ```suggestion
Spark Connect adopts proto3, which does not support `required` constraint anymore. 
```

> Connect adopts proto3, which does not support `required` constraint anymore. 

Reviewer 2: ```suggestion
Spark Connect adopts proto3, which does not support the use of the `required` constraint anymore. 
```

For non-message proto fields, there is also no `has_field_name` functions to easy tell
if a filed is set or not-set. (Read [proto3 field rules](https://developers.google.com/protocol-buffers/docs/proto3#specifying_field_rules))

### Required field

> Even proto3 does not offer `required` constraint, there are still some fields that
> are semantically required. For such case, we shall add comment `(Required)` on the
> field to indicate it as required. The expectation for client implementation 
> (or any submitted proto plan) is that such fields should be always set, and server will
> always assume such fields are set and use whatever values from the fields directly.
> It is the client side's fault to not offer meaningful value for `required` field and in that case,
> the behavior on the server side is decided by the default value of the field.

Reviewer 2: ```suggestion
When adding fields that have required semantics, developers are required to follow
the outlined process. Fields that are semantically required for the server to 
correctly process the incoming message must be documented with `(Required)`. For scalar
fields the server will not perform any additional input validation. For compound fields,
the server will perform minimal checks to avoid null pointer exceptions but will not
perform any semantic validation.
```

Example:
```protobuf
message DataSource {
 // (Required) Supported formats include: parquet, orc, text, json, parquet, csv, avro.
 string format = 1;
}
```

> ### Optional field and default value

Reviewer 2: ```suggestion
### Optional fields
```

> Semantically optional fields should be marked by proto3 `optional`. The server side will decide
> if to use the generated `has_field_name` to tell the field is set (when its default value
> is different from the Spark parameter default value) or use the field default value directly
> (when its default value is the same as Spark parameter default value). It is also required
> to use `(Optional)` in the comment to indicate this field is optional.

Reviewer 2: ```suggestion
Semantically optional fields must be marked by `optional`. The server side will
then use this information to branch into different behaviors based on the presence or absence of this field. 

Due to the lack of configurable default values for scalar types, the pure presence of
an optional value does not define its default value. The server side implementation will interpret the observed value based on its own rules.
```

# Required, Optional and default values

Connect adopts proto3, which does not support `required` constraint anymore.
Every field is optional. For non-message proto fields, there is also no `hasXXX`
functions to easy tell if a filed is set or not-set. For those non-message proto,

> they also have default values. E.g. int has 0 as the default value.

Reviewer 2: Hey, please have a look at https://developers.google.com/protocol-buffers/docs/proto3#specifying_field_rules and check if this is not applicable in this case.

### Required field

Even proto3 does not offer `required` constraint, there is still some fields that
are semantically required. For such case, we share add comment on the field to mark
it as required. The expectation for client implementation (or any submitted proto plan)
is that such fields should be always set, and server will always assume such fields
are set and use whatever values of the fields directly. It is the client side's fault
to not offer meaningful value for `required` field and in that case, the behavior on
the server side is not defined.

### Optional field and default value

Optional fields could have default values when the field is not set, and we are seeing
two cases that depends on whether the default value for the proto field is consistent
with the corresponding Spark parameter default value.

If both default values of the proto field and Spark parameter are the same, we keep the
proto field as is.

For example:
```Scala
// Spark Plan.
class FooPlan(size: Long = 0)
```

```protobuf
// Connect proto.
message Foo {

>   int64 size; 

Author: I intentionally do not add the proto field number.  This is to avoid people mis-understand it as a default value thus causing confusions.
